src/Firestore/WriteBatch.php

```diff
@@ -0,0 +1,545 @@
+     *           `['exists' => true]` (i.e. Document must exist in Firestore).
+     *           For more information, refer to the [Precondition](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1beta1#google.firestore.v1beta1.Precondition)
+     *           documentation.
+     * }
+     * @return WriteBatch
+     * @codingStandardsIgnoreEnd
+     */
+    public function update($documentName, array $fields, array $options = [])
+    {
+        if ($this->hasPreviousTransform) {
+            throw new \BadMethodCallException(
+                'Cannot apply an UPDATE operation after a TRANSFORM operation has been enqueued.'
+            );
+        }
+
+        $options += [
+            'precondition' => ['exists' => true]
+        ];
+
+        list($fields, $timestamps, $deletes) = $this->valueMapper->findSentinels($fields);
+
+        // We only want to enqueue an update write if there are non-sentinel fields
+        // OR no timestamp sentinels are found.
+        // We MUST always enqueue at least one write, so if there are no fields
+        // and no timestamp sentinels, we can assume an empty write is intended
+        // and enqueue an empty UPDATE operation.
+        if ($fields || !$timestamps) {
+            $this->writes[] = $this->createDatabaseWrite(self::TYPE_UPDATE, $documentName, [
+                'fields' => $this->valueMapper->encodeValues(
+                    $this->valueMapper->decodeFieldPaths($fields)
+                ),
+                'updateMask' => array_merge($this->valueMapper->encodeFieldPaths($fields), $deletes)
+            ] + $options);
+        }
+
+        // Setting values to the server timestamp is implemented as a document tranformation.
+        $this->updateTransformations($documentName, $timestamps);
+
+        return $this;
+    }
+
+    /**
+     * Enqueue a set mutation.
+     *
+     * Replaces all fields in a Firestore document.
+     *
+     * Example:
+     * ```
+     * $batch->set($documentName, [
+     *     'name' => 'John'
+     * ]);
+     *
+     * @codingStandardsIgnoreStart
+     * @param string $documentName The document to update.
+     * @param array $fields An array containing fields, where keys are the field
+     *        names, and values are field values. Nested arrays are allowed.
+     *        Note that unlike {@see Google\Cloud\Firestore\WriteBatch::update()},
+     *        field paths are NOT supported by this method.
+     * @param array $options {
+     *     Configuration Options
+     *
+     *     @type bool $merge If true, unwritten fields will be preserved.
+     *           Otherwise, they will be overwritten (removed). **Defaults to**
+     *           `false`.
+     * }
+     * @return WriteBatch
+     * @codingStandardsIgnoreEnd
+     */
+    public function set($documentName, array $fields, array $options = [])
+    {
+        if ($this->hasPreviousTransform) {
+            throw new \BadMethodCallException(
```

Comment:
It's ok to let the backend do this verification. This way, they can lift this restriction later.

Furthermore, this only applies on a per-document basis.